File: 678-valid-parenthesis-string/valid-parenthesis-string.java
import java.util.*;

class Solution {
    public boolean checkValidString(String s) {

        Stack<Integer> left = new Stack<>();
        Stack<Integer> star = new Stack<>();

        for (int i = 0; i < s.length(); i++) {

            char ch = s.charAt(i);

            if (ch == '(') {

                left.push(i);

            } else if (ch == '*') {

                star.push(i);

            } else {

                if (!left.isEmpty()) {
                    left.pop();
                }

                else if (!star.isEmpty()) {
                    star.pop();
                }

                else {
                    return false;
                }
            }
        }

        while (!left.isEmpty() && !star.isEmpty()) {

            if (left.peek() > star.peek()) {
                return false;
            }

            left.pop();
            star.pop();
        }

        return left.isEmpty();
    }
}

File: 678-valid-parenthesis-string/valid-parenthesis-string.py
class Solution:
    def checkValidString(self, s: str) -> bool:
        left = []
        star = []

        for i in range(len(s)):

            if s[i] == '(':
                left.append(i)

            elif s[i] == '*':
                star.append(i)

            else:
                if left:
                    left.pop()

                elif star:
                    star.pop()

                else:
                    return False

        while left and star:

            # Star must come after '('
            if left[-1] > star[-1]:
                return False

            left.pop()
            star.pop()

        return not left